extensions/amp-a4a/0.1/crypto-verifier.js

```diff
@@ -14,110 +14,89 @@
 
+/** @const {?SubtleCrypto} */
+const crossCrypto = window.crypto ?
+    (isWebkit ? window.crypto.webkitSubtle : window.crypto.subtle) : null;
+
+/** @const {number} */
 const VERSION = 0x00;
 
 /**
- * An object holding the public key, and stuff derived from it.
+ * An object holding the public key and its hash.
  *
  * @typedef {{
- *   publicKey: !Object,
- *   hash: Uint8Array,
- *   cryptoKey: CryptoKey
+ *   hash: !Uint8Array,
+ *   cryptoKey: !CryptoKey
  * }}
  */
 let PublicKeyInfoDef;
 
 /**
- * Compute and cache hash and CryptoKey of public key
- * @param {!Object} publicKey Parsed JSON web key
+ * Convert a JSON Web Key object to a browser-native cryptographic key and
+ * compute a hash for it.  The caller must verify that Web Cryptography is
+ * available using isCryptoAvailable before calling this function.
+ *
+ * @param {!Object} jwk An object which is hopefully an RSA JSON Web Key.  The
+ *     caller should verify that it is an object before calling this function.
  * @return {!Promise<!PublicKeyInfoDef>}
  */
-export function importPublicKey(publicKey) {
-  const lenMod = lenPrefix(base64UrlDecodeToBytes(publicKey['n']));
-  const lenPubExp = lenPrefix(base64UrlDecodeToBytes(publicKey['e']));
-  const data = new Uint8Array(lenMod.length + lenPubExp.length);
-  data.set(lenMod);
-  data.set(lenPubExp, lenMod.length);
-  return crossCrypto.digest({
-    // The list of RSA public keys are not under attacker's control,
-    // so a collision would not help.
-    name: 'SHA-1',
-  }, data)
-    .then(digest => {
-      // Hash is the first 4 bytes of the SHA-1 digest.
-      const hash = new Uint8Array(digest, 0, 4);
-
-      // Now Get the CryptoKey.
-      const jsonPublicKey = isWebkit ?
-            // Webkit wants this as an ArrayBuffer.
-            stringToBytes(JSON.stringify(publicKey)) :
-            publicKey;
-      // Convert the key to internal CryptoKey format.
-      return crossCrypto.importKey(
-        'jwk',
-        jsonPublicKey,
-        {
-          name: 'RSASSA-PKCS1-v1_5',
-          hash: {name: 'SHA-256'},
-        },
-        true,
-        ['verify'])
-        .then(cryptoKey => ({publicKey, hash, cryptoKey}));
-    });
+export function importPublicKey(jwk) {
+  // WebKit wants this as an ArrayBufferView.
+  return (isWebkit ? utf8Encode(JSON.stringify(jwk)) : Promise.resolve(jwk))
+      .then(encodedJwk =>
+      crossCrypto.importKey(
+          'jwk',
+          encodedJwk,
+          {name: 'RSASSA-PKCS1-v1_5', hash: {name: 'SHA-256'}},
+          true,
+          ['verify']))
+              .then(cryptoKey => {
+                // We do the importKey first to allow the browser to check for
+                // an invalid key.  This last check is in case the key is valid
+                // but a different kind.
+                if (typeof jwk.n != 'string' || typeof jwk.e != 'string') {
+                  throw new Error('missing fields in JSON Web Key');
```

Comment:
Sorry for the back-and-forth it might have caused. But just want to make it clear of the common practices here so that we can follow. 

@jridgewell why this is an anti-pattern? I thought it's more natural to use Promise's own way of error handling (reject) in a Promise block, instead of using a throw. 

Meantime, I do see return a rejection in then, for example [here](https://github.com/ampproject/amphtml/blob/fa84fd2a584ffec47510cd0440a4cc0e7a40ecd3/extensions/amp-fx-flying-carpet/0.1/test/test-amp-fx-flying-carpet.js#L70).